File: src/algorithms/arrays/sliding-window/min-size-subarray-sum/min-size-subarray-sum_test.go
package minsizesubarraysum

import "testing"

func TestMinSizeSubarraySumBasic(t *testing.T) {
	minLength, startIndex := minSizeSubarraySum([]int{2, 3, 1, 2, 4, 3}, 7)
	if minLength != 2 || startIndex != 4 {
		t.Errorf("expected (2, 4), got (%d, %d)", minLength, startIndex)
	}
}

func TestMinSizeSubarraySumSingleElementMeetsTarget(t *testing.T) {
	minLength, _ := minSizeSubarraySum([]int{1, 4, 4}, 4)
	if minLength != 1 {
		t.Errorf("expected 1, got %d", minLength)
	}
}

func TestMinSizeSubarraySumNoValidSubarray(t *testing.T) {
	minLength, _ := minSizeSubarraySum([]int{1, 1, 1, 1}, 10)
	if minLength != 0 {
		t.Errorf("expected 0, got %d", minLength)
	}
}

func TestMinSizeSubarraySumWholeArray(t *testing.T) {
	minLength, startIndex := minSizeSubarraySum([]int{1, 2, 3}, 6)
	if minLength != 3 || startIndex != 0 {
		t.Errorf("expected (3, 0), got (%d, %d)", minLength, startIndex)
	}
}

func TestMinSizeSubarraySumEmptyArray(t *testing.T) {
	minLength, _ := minSizeSubarraySum([]int{}, 7)
	if minLength != 0 {
		t.Errorf("expected 0, got %d", minLength)
	}
}

func TestMinSizeSubarraySumZeroTarget(t *testing.T) {
	minLength, _ := minSizeSubarraySum([]int{1, 2, 3}, 0)
	if minLength != 0 {
		t.Errorf("expected 0, got %d", minLength)
	}
}

func TestMinSizeSubarraySumSingleElementExact(t *testing.T) {
	minLength, startIndex := minSizeSubarraySum([]int{7}, 7)
	if minLength != 1 || startIndex != 0 {
		t.Errorf("expected (1, 0), got (%d, %d)", minLength, startIndex)
	}
}

func TestMinSizeSubarraySumRepeatedElements(t *testing.T) {
	minLength, _ := minSizeSubarraySum([]int{3, 3, 3, 3}, 6)
	if minLength != 2 {
		t.Errorf("expected 2, got %d", minLength)
	}
}

func TestMinSizeSubarraySumLargeFirstElement(t *testing.T) {
	minLength, startIndex := minSizeSubarraySum([]int{100, 1, 1, 1, 1}, 100)
	if minLength != 1 || startIndex != 0 {
		t.Errorf("expected (1, 0), got (%d, %d)", minLength, startIndex)
	}
}

File: src/algorithms/arrays/sliding-window/min-size-subarray-sum/__tests__/min-size-subarray-sum_test.rs
include!("../sources/min-size-subarray-sum.rs");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_min_size_subarray_sum_basic() {
        let (min_length, start_index) = min_size_subarray_sum(&[2, 3, 1, 2, 4, 3], 7);
        assert_eq!(min_length, 2);
        assert_eq!(start_index, 4);
    }

    #[test]
    fn test_min_size_subarray_sum_single_element_meets_target() {
        let (min_length, _) = min_size_subarray_sum(&[1, 4, 4], 4);
        assert_eq!(min_length, 1);
    }

    #[test]
    fn test_min_size_subarray_sum_no_valid_subarray() {
        let (min_length, _) = min_size_subarray_sum(&[1, 1, 1, 1], 10);
        assert_eq!(min_length, 0);
    }

    #[test]
    fn test_min_size_subarray_sum_whole_array() {
        let (min_length, start_index) = min_size_subarray_sum(&[1, 2, 3], 6);
        assert_eq!(min_length, 3);
        assert_eq!(start_index, 0);
    }

    #[test]
    fn test_min_size_subarray_sum_empty_array() {
        let (min_length, _) = min_size_subarray_sum(&[], 7);
        assert_eq!(min_length, 0);
    }

    #[test]
    fn test_min_size_subarray_sum_zero_target() {
        let (min_length, _) = min_size_subarray_sum(&[1, 2, 3], 0);
        assert_eq!(min_length, 0);
    }

    #[test]
    fn test_min_size_subarray_sum_single_element_exact() {
        let (min_length, start_index) = min_size_subarray_sum(&[7], 7);
        assert_eq!(min_length, 1);
        assert_eq!(start_index, 0);
    }

    #[test]
    fn test_min_size_subarray_sum_repeated_elements() {
        let (min_length, _) = min_size_subarray_sum(&[3, 3, 3, 3], 6);
        assert_eq!(min_length, 2);
    }

    #[test]
    fn test_min_size_subarray_sum_large_first_element() {
        let (min_length, start_index) = min_size_subarray_sum(&[100, 1, 1, 1, 1], 100);
        assert_eq!(min_length, 1);
        assert_eq!(start_index, 0);
    }
}
